File: platform/backend/services/email_service.py
"""

Email Service for Iacgenie AI

Handles transactional email sending via SendGrid with templates and delivery tracking

Features:

- SendGrid integration for email delivery

- HTML email templates using Jinja2

- Retry logic with exponential backoff

- Delivery status tracking in database

"""

import os

import logging

from typing import Dict, Any, Optional

from datetime import datetime

from dataclasses import dataclass, field

logger = logging.getLogger(__name__)

# Try to import SendGrid

try:
    from sendgrid import SendGridAPIClient
    from sendgrid.helpers.mail import (
        Mail,
        From,
        To,
        Subject,
        PlainTextContent,
        HTMLContent,
        MailSettings,
        SandBoxMode,
    )

    SENDGRID_AVAILABLE = True
except ImportError:
    logger.warning("SendGrid not installed. Install with: pip install sendgrid")
    SENDGRID_AVAILABLE = False
try:
    from jinja2 import Template
except ImportError:
    logger.warning("Jinja2 not installed. Install with: pip install jinja2")
    _Template: Any = None  # type: ignore[misc,assignment,no-redef]
    Template: Any = None  # type: ignore[misc,assignment,no-redef]


@dataclass
class EmailDeliveryResult:
    """Result of email delivery attempt"""

    success: bool
    message_id: Optional[str] = None
    error_message: Optional[str] = None
    status_code: int = 0
    timestamp: datetime = field(default_factory=datetime.utcnow)
    retry_count: int = 0


class EmailService:
    """
    Email service using SendGrid for transactional emails
    Features:
    - HTML email templates
    - Retry logic (max 3 retries with exponential backoff)
    - Sandboxing for development
    - Delivery tracking
    """

    def __init__(self) -> None:
        self.api_key = os.getenv("SENDGRID_API_KEY", "")
        self.from_email = os.getenv("EMAIL_FROM_ADDRESS", "noreply@iacgenie.ai")
        self.is_sandbox = os.getenv("SENDGRID_SANDBOX", "false").lower() == "true"
        self.max_retries = int(os.getenv("EMAIL_MAX_RETRIES", "3"))
        self.base_retry_delay = float(os.getenv("EMAIL_RETRY_DELAY", "1.0"))
        self._client: Optional[SendGridAPIClient] = None
        if SENDGRID_AVAILABLE and self.api_key:
            try:
                self._client = SendGridAPIClient(self.api_key)
                logger.info(f"SendGrid initialized (sandbox={self.is_sandbox})")
            except Exception as e:
                logger.error(f"Failed to initialize SendGrid: {e}")
        # Email templates (will be loaded from files or fallback to inline)
        self._templates: Dict[str, str] = {}
        self._load_templates()

    def _load_templates(self) -> None:
        """Load email templates from environment or use inline defaults"""
        # Check for template file paths
        self.verification_template_path = os.getenv(
            "EMAIL_VERIFICATION_TEMPLATE_PATH",
            "backend/templates/verification_email.html",
        )
        self.password_reset_template_path = os.getenv(
            "EMAIL_PASSWORD_RESET_TEMPLATE_PATH",
            "backend/templates/password_reset_email.html",
        )
        self.invitation_template_path = os.getenv(
            "EMAIL_INVITATION_TEMPLATE_PATH", "backend/templates/invitation_email.html"
        )
        # Inline templates as fallback
        ver_t = (
            "<!DOCTYPE html>\n"
            "<html>\n"
            "<head>\n"
            '    <meta charset="UTF-8">\n'
            "    <title>Verify Your Email - Iacgenie AI</title>\n"
            "</head>\n"
            '<body style="font-family: Arial, sans-serif; '
            "line-height: 1.6; color: #333; "
            "max-width: 600px; margin: 0 auto; "
            'padding: 20px;">\n'
            '    <div style="background: '
            "linear-gradient(135deg, #f6d365 0%, "
            "#fda085 100%); padding: 30px; "
            "text-align: center; "
            'border-radius: 10px 10px 0 0;">\n'
            '        <h1 style="color: white; margin: 0;">'
            "Iacgenie AI</h1>\n"
            '        <p style="color: white; margin: 5px 0 0;">'
            "Infrastructure as Code Generation Platform</p>\n"
            "    </div>\n"
            '    <div style="background: #f9f9f9; padding: 30px; '
            'border-radius: 0 0 10px 10px;">\n'
            "        <h2>Verify Your Email Address</h2>\n"
            "        <p>Hello {{ user_name or 'there' }},</p>\n"
            "        <p>Thank you for signing up with "
            "Iacgenie AI! Please verify your email address "
            "to complete your registration and get started "
            "with our platform.</p>\n"
            '        <div style="text-align: center; '
            'margin: 30px 0;">\n'
            '            <a href="{{ verification_url }}" '
            'style="background: linear-gradient(135deg, '
            "#f6d365 0%, #fda085 100%); color: white; "
            "padding: 15px 30px; text-decoration: none; "
            'border-radius: 5px; font-weight: bold;">'
            "Verify Email Address</a>\n"
            "        </div>\n"
            '        <p style="font-size: 14px; color: #666;">'
            "Or copy and paste this link into your browser:</p>\n"
            '        <p style="font-size: 12px; color: #999; '
            'word-break: break-all;">'
            "{{ verification_url }}</p>\n"
            '        <hr style="border: none; '
            'border-top: 1px solid #ddd; margin: 30px 0;">\n'
            '        <p style="font-size: 12px; color: #999;">\n'
            "            If you didn't create an account with "
            "Iacgenie AI, please ignore this email.<br>\n"
            "            This verification link will expire "
            "in 24 hours.\n"
            "        </p>\n"
            "    </div>\n"
            '    <div style="text-align: center; padding: 20px; '
            'color: #999;">\n'
            "        <p>Iacgenie AI - Making Infrastructure "
            "Code Easy</p>\n"
            "    </div>\n"
            "</body>\n"
            "</html>\n"
        )
        pwd_t = (
            "<!DOCTYPE html>\n"
            "<html>\n"
            "<head>\n"
            '    <meta charset="UTF-8">\n'
            "    <title>Reset Your Password - Iacgenie AI</title>\n"
            "</head>\n"
            '<body style="font-family: Arial, sans-serif; '
            "line-height: 1.6; color: #333; "
            "max-width: 600px; margin: 0 auto; "
            'padding: 20px;">\n'
            '    <div style="background: '
            "linear-gradient(135deg, #f6d365 0%, "
            "#fda085 100%); padding: 30px; "
            "text-align: center; "
            'border-radius: 10px 10px 0 0;">\n'
            '        <h1 style="color: white; margin: 0;">'
            "Iacgenie AI</h1>\n"
            '        <p style="color: white; margin: 5px 0 0;">'
            "Infrastructure as Code Generation Platform</p>\n"
            "    </div>\n"
            '    <div style="background: #f9f9f9; padding: 30px; '
            'border-radius: 0 0 10px 10px;">\n'
            "        <h2>Reset Your Password</h2>\n"
            "        <p>Hello {{ user_name or 'there' }},</p>\n"
            "        <p>We received a request to reset your "
            "password for your Iacgenie AI account.</p>\n"
            '        <div style="text-align: center; '
            'margin: 30px 0;">\n'
            '            <a href="{{ reset_url }}" '
            'style="background: linear-gradient(135deg, '
            "#f6d365 0%, #fda085 100%); color: white; "
            "padding: 15px 30px; text-decoration: none; "
            'border-radius: 5px; font-weight: bold;">'
            "Reset Password</a>\n"
            "        </div>\n"
            '        <p style="font-size: 14px; color: #666;">'
            "Or copy and paste this link into your browser:</p>\n"
            '        <p style="font-size: 12px; color: #999; '
            'word-break: break-all;">{{ reset_url }}</p>\n'
            '        <div style="background: #fff3cd; padding: '
            '15px; border-radius: 5px; margin-top: 20px;">\n'
            '            <p style="color: #856404; margin: 0;">'
            "This link will expire in 2 hours.</p>\n"
            "        </div>\n"
            '        <hr style="border: none; '
            'border-top: 1px solid #ddd; margin: 30px 0;">\n'
            '        <p style="font-size: 12px; color: #999;">\n'
            "            If you didn't request a password reset, "
            "please ignore this email.<br>\n"
            "            Your current password will remain "
            "unchanged.\n"
            "        </p>\n"
            "    </div>\n"
            '    <div style="text-align: center; padding: 20px; '
            'color: #999;">\n'
            "        <p>Iacgenie AI - Making Infrastructure "
            "Code Easy</p>\n"
            "    </div>\n"
            "</body>\n"
            "</html>\n"
        )
        inv_t = (
            "<!DOCTYPE html>\n"
            "<html>\n"
            "<head>\n"
            '    <meta charset="UTF-8">\n'
            "    <title>You've Been Invited - Iacgenie AI</title>\n"
            "</head>\n"
            '<body style="font-family: Arial, sans-serif; '
            "line-height: 1.6; color: #333; "
            "max-width: 600px; margin: 0 auto; "
            'padding: 20px;">\n'
            '    <div style="background: '
            "linear-gradient(135deg, #f6d365 0%, "
            "#fda085 100%); padding: 30px; "
            "text-align: center; "
            'border-radius: 10px 10px 0 0;">\n'
            '        <h1 style="color: white; margin: 0;">'
            "Iacgenie AI</h1>\n"
            '        <p style="color: white; margin: 5px 0 0;">'
            "Infrastructure as Code Generation Platform</p>\n"
            "    </div>\n"
            '    <div style="background: #f9f9f9; padding: 30px; '
            'border-radius: 0 0 10px 10px;">\n'
            "        <h2>You've Been Invited to Join "
            "Iacgenie AI</h2>\n"
            "        <p>Hello {{ user_name or 'there' }},</p>\n"
            "        <p><strong>{{ inviter_name }}</strong> "
            "has invited you to join their Iacgenie AI "
            "team with <strong>{{ role }}</strong> access.</p>\n"
            '        <div style="text-align: center; '
            'margin: 30px 0;">\n'
            '            <a href="{{ invitation_url }}" '
            'style="background: linear-gradient(135deg, '
            "#f6d365 0%, #fda085 100%); color: white; "
            "padding: 15px 30px; text-decoration: none; "
            'border-radius: 5px; font-weight: bold;">'
            "Join Team</a>\n"
            "        </div>\n"
            '        <p style="font-size: 14px; color: #666;">'
            "Or copy and paste this link into your browser:</p>\n"
            '        <p style="font-size: 12px; color: #999; '
            'word-break: break-all;">'
            "{{ invitation_url }}</p>\n"
            '        <div style="background: #d1ecf1; padding: '
            '15px; border-radius: 5px; margin-top: 20px;">\n'
            '            <p style="color: #0c5460; margin: 0;">'
            "Create your Iacgenie AI account to complete your "
            "invitation.</p>\n"
            "        </div>\n"
            '        <hr style="border: none; '
            'border-top: 1px solid #ddd; margin: 30px 0;">\n'
            '        <p style="font-size: 12px; color: #999;">\n'
            "            This invitation will expire in 7 days.<br>\n"
            "            If you didn't expect to receive this "
            "invitation, please ignore this email.\n"
            "        </p>\n"
            "    </div>\n"
            '    <div style="text-align: center; padding: 20px; '
            'color: #999;">\n'
            "        <p>Iacgenie AI - Making Infrastructure "
            "Code Easy</p>\n"
            "    </div>\n"
            "</body>\n"
            "</html>\n"
        )
        self._inline_templates = {
            "verification": ver_t,
            "password_reset": pwd_t,
            "invitation": inv_t,
        }

    def _get_template(self, template_name: str) -> str:
        """Get email template content"""
        if template_name in self._templates:
            return self._templates[template_name]
        # Try to load from file first
        template_path = None
        if template_name == "verification":
            template_path = self.verification_template_path
        elif template_name == "password_reset":
            template_path = self.password_reset_template_path
        elif template_name == "invitation":
            template_path = self.invitation_template_path
        if template_path:
            try:
                with open(template_path, "r") as f:
                    content = f.read()
                    self._templates[template_name] = content
                    return content
            except FileNotFoundError:
                logger.warning(
                    f"Template file not found: {template_path}. Using inline template."
                )
            except Exception as e:
                logger.error(f"Failed to load template {template_path}: {e}")
        # Return inline template
        return self._inline_templates.get(template_name, "")

    def _render_template(self, template_name: str, context: Dict[str, Any]) -> str:
        """Render email template with context"""
        if Template is None:
            logger.error("Jinja2 is not installed")
            return "Please install Jinja2: pip install jinja2"
        template_content = self._get_template(template_name)
        template = Template(template_content)
        # nosemgrep: python.flask.security.xss.audit.direct-use-of-jinja2.direct-use-of-jinja2
        return template.render(**context)

    async def _send_email_direct(
        self, to_email: str, subject: str, html_content: str
    ) -> EmailDeliveryResult:
        """Send email directly via SendGrid API"""
        if not self._client:
            logger.error("SendGrid client not initialized")
            return EmailDeliveryResult(
                success=False, error_message="SendGrid API key not configured"
            )
        try:
            message = Mail(
                from_email=From(self.from_email),
                to_emails=To(to_email),
                subject=Subject(subject),
                plain_text_content=PlainTextContent(self._strip_html(html_content)),
                html_content=HTMLContent(html_content),
            )
            # Enable sandbox mode for testing
            if self.is_sandbox:
                message.mail_settings = MailSettings(
                    sandbox_mode=SandBoxMode(enable=True)
                )
            response = self._client.send(message)
            return EmailDeliveryResult(
                success=response.status_code == 202,
                message_id=response.headers.get("x-message-id"),
                status_code=response.status_code,
            )
        except Exception as e:
            logger.error(f"Failed to send email: {e}")
            return EmailDeliveryResult(success=False, error_message=str(e))

    def _strip_html(self, html: str) -> str:
        """Strip HTML tags from content for plain text version"""
        import re

        clean = re.compile("<.*?>")
        return re.sub(clean, "", html)

    async def send_verification_email(
        self, to_email: str, verification_url: str, user_name: Optional[str] = None
    ) -> EmailDeliveryResult:
        """Send email verification link"""
        if not self.api_key:
            logger.warning(
                f"SENDGRID_API_KEY not configured. "
                f"Would send verification to {to_email}"
            )
            return EmailDeliveryResult(
                success=True,
                error_message="Email service not configured (sandbox mode)",
            )
        subject = "Verify Your Email - Iacgenie AI"
        html_content = self._render_template(
            "verification",
            {"user_name": user_name, "verification_url": verification_url},
        )
        return await self._send_email_with_retry(to_email, subject, html_content)

    async def send_password_reset_email(
        self, to_email: str, reset_url: str, user_name: Optional[str] = None
    ) -> EmailDeliveryResult:
        """Send password reset link"""
        if not self.api_key:
            logger.warning(
                f"SENDGRID_API_KEY not configured. Would send reset email to {to_email}"
            )
            return EmailDeliveryResult(
                success=True,
                error_message="Email service not configured (sandbox mode)",
            )
        subject = "Reset Your Password - Iacgenie AI"
        html_content = self._render_template(
            "password_reset", {"user_name": user_name, "reset_url": reset_url}
        )
        return await self._send_email_with_retry(to_email, subject, html_content)

    async def send_invitation_email(
        self,
        to_email: str,
        invitation_url: str,
        user_name: Optional[str] = None,
        inviter_name: Optional[str] = None,
        role: str = "user",
    ) -> EmailDeliveryResult:
        """Send team invitation email"""
        if not self.api_key:
            logger.warning(
                f"SENDGRID_API_KEY not configured. Would send invitation to {to_email}"
            )
            return EmailDeliveryResult(
                success=True,
                error_message="Email service not configured (sandbox mode)",
            )
        subject = "You've Been Invited to Iacgenie AI"
        html_content = self._render_template(
            "invitation",
            {
                "user_name": user_name,
                "inviter_name": inviter_name or "A team member",
                "invitation_url": invitation_url,
                "role": role,
            },
        )
        return await self._send_email_with_retry(to_email, subject, html_content)

    async def _send_email_with_retry(
        self, to_email: str, subject: str, html_content: str
    ) -> EmailDeliveryResult:
        """Send email with retry logic"""
        last_result = None
        for attempt in range(self.max_retries + 1):
            result = await self._send_email_direct(to_email, subject, html_content)
            if result.success:
                logger.info(f"Email sent successfully to {to_email}")
                return result
            last_result = result
            logger.warning(
                f"Email attempt {attempt + 1} failed: {result.error_message}"
            )
            if attempt < self.max_retries:
                delay = self.base_retry_delay * (2**attempt)
                import asyncio

                logger.info(f"Retrying in {delay} seconds...")
                await asyncio.sleep(delay)
        logger.error(f"All email delivery attempts failed for {to_email}")
        return last_result  # type: ignore

    async def send_email(
        self, to_email: str, subject: str, html_content: str
    ) -> EmailDeliveryResult:
        """Generic email sending method"""
        if not self.api_key:
            logger.warning(f"SENDGRID_API_KEY not configured. Would send to {to_email}")
            return EmailDeliveryResult(
                success=True, error_message="Email service not configured"
            )
        return await self._send_email_with_retry(to_email, subject, html_content)

    def is_configured(self) -> bool:
        """Check if email service is properly configured"""
        return bool(self.api_key and self._client)

    def get_provider_name(self) -> str:
        """Get email provider name"""
        return "SendGrid" if SENDGRID_AVAILABLE else "Not Configured"


# Global instance


email_service = EmailService()


async def get_email_service() -> EmailService:
    """Get email service instance"""
    global email_service
    if not hasattr(email_service, "initialized"):
        email_service = EmailService()
    return email_service

File: lightserv/src/email-service.ts
/**
 * LightSerp Email Service — SMTP2GO integration
 *
 * Handles transactional email sending via SMTP2GO API with HTML templates
 * for verification, password reset, and invitation flows.
 *
 * Adapted from TerraGenius iacgenie backend.
 *
 * Environment variables:
 *   SMTP2GO_API_KEY          SMTP2GO API key
 *   EMAIL_FROM_ADDRESS       Sender email address
 *   SMTP_SERVER              SMTP2GO server (default: mail.smtp2go.com)
 *   SMTP_PORT                SMTP2GO port (default: 2525)
 *   EMAIL_MAX_RETRIES        Max send retries (default: 3)
 *   EMAIL_RETRY_DELAY        Base retry delay in seconds (default: 1.0)
 *   SMTP2GO_SANDBOX          Enable sandbox mode (default: false)
 *   LIGHTSERP_URL            Frontend base URL for generating links
 */

import https from 'node:https';
import { log } from './logger.js';

// ── Configuration ─────────────────────────────────────────────────────

const SMTP2GO_API_KEY = process.env.SMTP2GO_API_KEY || '';
const EMAIL_FROM_ADDRESS = process.env.EMAIL_FROM_ADDRESS || 'noreply@lightserp.ai';
const SMTP_SERVER = process.env.SMTP_SERVER || 'mail.smtp2go.com';
const SMTP_PORT = parseInt(process.env.SMTP_PORT || '2525', 10);
const EMAIL_MAX_RETRIES = parseInt(process.env.EMAIL_MAX_RETRIES || '3', 10);
const EMAIL_RETRY_DELAY = parseFloat(process.env.EMAIL_RETRY_DELAY || '1.0');
const LIGHTSERP_URL = process.env.LIGHTSERP_URL || 'http://localhost:3071';

// ── Email Delivery Result ─────────────────────────────────────────────

interface EmailDeliveryResult {
  success: boolean;
  messageId?: string;
  errorMessage?: string;
  statusCode: number;
  timestamp: Date;
  retryCount: number;
}

// ── Email Templates (inline, brand-adapted to LightSerp) ──────────────

function getVerificationTemplate(
  verificationUrl: string,
  userName?: string
): string {
  const name = userName || 'there';
  return `<!DOCTYPE html>
<html>
<head>
  <meta charset="UTF-8">
  <title>Verify Your Email - LightSerp</title>
</head>
<body style="font-family: Arial, sans-serif; line-height: 1.6; color: #333; max-width: 600px; margin: 0 auto; padding: 20px;">
  <div style="background: linear-gradient(135deg, #667eea 0%, #764ba2 100%); padding: 30px; text-align: center; border-radius: 10px 10px 0 0;">
    <h1 style="color: white; margin: 0;">LightSerp</h1>
    <p style="color: white; margin: 5px 0 0;">Search &amp; Scrape MCP for AI Models</p>
  </div>
  <div style="background: #f9f9f9; padding: 30px; border-radius: 0 0 10px 10px;">
    <h2>Verify Your Email Address</h2>
    <p>Hello ${name},</p>
    <p>Thank you for signing up with LightSerp! Please verify your email address to complete your registration.</p>
    <div style="text-align: center; margin: 30px 0;">
      <a href="${verificationUrl}" style="background: linear-gradient(135deg, #667eea 0%, #764ba2 100%); color: white; padding: 15px 30px; text-decoration: none; border-radius: 5px; font-weight: bold;">Verify Email Address</a>
    </div>
    <p style="font-size: 14px; color: #666;">Or copy and paste this link into your browser:</p>
    <p style="font-size: 12px; color: #999; word-break: break-all;">${verificationUrl}</p>
    <hr style="border: none; border-top: 1px solid #ddd; margin: 30px 0;">
    <p style="font-size: 12px; color: #999;">
      If you didn't create an account with LightSerp, please ignore this email.<br>
      This verification link will expire in 24 hours.
    </p>
  </div>
  <div style="text-align: center; padding: 20px; color: #999;">
    <p>LightSerp - Search &amp; Scrape MCP for AI Models</p>
  </div>
</body>
</html>`;
}

function getPasswordResetTemplate(
  resetUrl: string,
  userName?: string
): string {
  const name = userName || 'there';
  return `<!DOCTYPE html>
<html>
<head>
  <meta charset="UTF-8">
  <title>Reset Your Password - LightSerp</title>
</head>
<body style="font-family: Arial, sans-serif; line-height: 1.6; color: #333; max-width: 600px; margin: 0 auto; padding: 20px;">
  <div style="background: linear-gradient(135deg, #667eea 0%, #764ba2 100%); padding: 30px; text-align: center; border-radius: 10px 10px 0 0;">
    <h1 style="color: white; margin: 0;">LightSerp</h1>
    <p style="color: white; margin: 5px 0 0;">Search &amp; Scrape MCP for AI Models</p>
  </div>
  <div style="background: #f9f9f9; padding: 30px; border-radius: 0 0 10px 10px;">
    <h2>Reset Your Password</h2>
    <p>Hello ${name},</p>
    <p>We received a request to reset your password for your LightSerp account.</p>
    <div style="text-align: center; margin: 30px 0;">
      <a href="${resetUrl}" style="background: linear-gradient(135deg, #667eea 0%, #764ba2 100%); color: white; padding: 15px 30px; text-decoration: none; border-radius: 5px; font-weight: bold;">Reset Password</a>
    </div>
    <p style="font-size: 14px; color: #666;">Or copy and paste this link into your browser:</p>
    <p style="font-size: 12px; color: #999; word-break: break-all;">${resetUrl}</p>
    <div style="background: #fff3cd; padding: 15px; border-radius: 5px; margin-top: 20px;">
      <p style="color: #856404; margin: 0;"><strong>This link will expire in 2 hours.</strong></p>
    </div>
    <hr style="border: none; border-top: 1px solid #ddd; margin: 30px 0;">
    <p style="font-size: 12px; color: #999;">
      If you didn't request a password reset, please ignore this email.<br>
      Your current password will remain unchanged.
    </p>
  </div>
  <div style="text-align: center; padding: 20px; color: #999;">
    <p>LightSerp - Search &amp; Scrape MCP for AI Models</p>
  </div>
</body>
</html>`;
}

function getWelcomeTemplate(userName: string, email: string): string {
  return `<!DOCTYPE html>
<html>
<head>
  <meta charset="UTF-8">
  <title>Welcome to LightSerp</title>
</head>
<body style="font-family: Arial, sans-serif; line-height: 1.6; color: #333; max-width: 600px; margin: 0 auto; padding: 20px;">
  <div style="background: linear-gradient(135deg, #667eea 0%, #764ba2 100%); padding: 30px; text-align: center; border-radius: 10px 10px 0 0;">
    <h1 style="color: white; margin: 0;">LightSerp</h1>
    <p style="color: white; margin: 5px 0 0;">Search &amp; Scrape MCP for AI Models</p>
  </div>
  <div style="background: #f9f9f9; padding: 30px; border-radius: 0 0 10px 10px;">
    <h2>Welcome to LightSerp, ${userName}!</h2>
    <p>Your account has been successfully created.</p>
    <p>You can now use LightSerp to power your AI agents with real-time search and web scraping capabilities.</p>
    <div style="text-align: center; margin: 30px 0;">
      <a href="${LIGHTSERP_URL}" style="background: linear-gradient(135deg, #667eea 0%, #764ba2 100%); color: white; padding: 15px 30px; text-decoration: none; border-radius: 5px; font-weight: bold;">Get Started</a>
    </div>
    <hr style="border: none; border-top: 1px solid #ddd; margin: 30px 0;">
    <p style="font-size: 12px; color: #999;">
      Account: ${email}<br>
      If you have any questions, feel free to reach out.
    </p>
  </div>
  <div style="text-align: center; padding: 20px; color: #999;">
    <p>LightSerp - Search &amp; Scrape MCP for AI Models</p>
  </div>
</body>
</html>`;
}

// ── SMTP2GO API Email Sending ────────────────────────────────────────

function smtp2goSend(
  to: string,
  subject: string,
  htmlBody: string
): Promise<EmailDeliveryResult> {
  return new Promise((resolve) => {
    const payload = JSON.stringify({
      to: [to],
      sender: EMAIL_FROM_ADDRESS,
      subject: subject,
      html_body: htmlBody,
    });

    const options = {
      hostname: SMTP_SERVER,
      port: SMTP_PORT,
      path: '/api/v3/email/send',
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'X-Smtp2go-Api-Key': SMTP2GO_API_KEY,
        'Content-Length': Buffer.byteLength(payload),
      },
    };

    const EMAIL_TIMEOUT = parseInt(process.env.EMAIL_TIMEOUT || '10000', 10);

    const req = https.request(options, (res) => {
      let data = '';
      res.on('data', (chunk) => { data += chunk; });
      res.on('end', () => {
        try {
          const result = JSON.parse(data);
          log.debug(`[SMTP2GO] Status ${res.statusCode}: ${JSON.stringify(result)}`);

          // SMTP2GO returns success with status 200 and data.succeeded > 0
          if (res.statusCode === 200 && result.data?.succeeded && result.data.succeeded > 0) {
            resolve({
              success: true,
              messageId: result.data?.messages?.[0]?.message_id,
              statusCode: res.statusCode ?? 0,
              timestamp: new Date(),
              retryCount: 0,
            });
          } else {
            const errMsg = result.message || result.errors?.[0] || `HTTP ${res.statusCode ?? 0}`;
            resolve({
              success: false,
              errorMessage: errMsg,
              statusCode: res.statusCode ?? 0,
              timestamp: new Date(),
              retryCount: 0,
            });
          }
        } catch {
          resolve({
            success: false,
            errorMessage: `Failed to parse SMTP2GO response: ${data}`,
            statusCode: res.statusCode ?? 0,
            timestamp: new Date(),
            retryCount: 0,
          });
        }
      });
    });

    req.on('error', (err) => {
      log.error(`[SMTP2GO] Request failed: ${err.message}`);
      resolve({
        success: false,
        errorMessage: err.message,
        statusCode: 0,
        timestamp: new Date(),
        retryCount: 0,
      });
    });

    req.on('timeout', () => {
      log.warn('[SMTP2GO] Request timed out after ' + EMAIL_TIMEOUT + 'ms');
      req.destroy();
      resolve({
        success: false,
        errorMessage: 'Request timed out',
        statusCode: 0,
        timestamp: new Date(),
        retryCount: 0,
      });
    });
    req.setTimeout(EMAIL_TIMEOUT);

    req.write(payload);
    req.end();
  });
}

// ── Email Validation ─────────────────────────────────────────────────

function validateEmail(email: string): { valid: boolean; error?: string } {
  if (!email || !email.trim()) {
    return { valid: false, error: 'Email is required' };
  }
  const trimmed = email.trim();
  const pattern = /^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$/;
  if (!pattern.test(trimmed)) {
    return { valid: false, error: `Invalid email format: ${trimmed}` };
  }
  if (trimmed.length > 254) {
    return { valid: false, error: 'Email address too long' };
  }
  if (trimmed.length < 3) {
    return { valid: false, error: 'Email address too short' };
  }
  return { valid: true };
}

// ── Retry Logic ──────────────────────────────────────────────────────

async function sendWithRetry(
  toEmail: string,
  subject: string,
  htmlContent: string,
  attempt = 0
): Promise<EmailDeliveryResult> {
  const result = await smtp2goSend(toEmail, subject, htmlContent);
  result.retryCount = attempt + 1;

  if (result.success) {
    log.info(`Email sent successfully to ${toEmail}`);
    return result;
  }

  log.warn(`Email attempt ${attempt + 1} failed: ${result.errorMessage}`);

  if (attempt < EMAIL_MAX_RETRIES) {
    const delay = EMAIL_RETRY_DELAY * Math.pow(2, attempt);
    log.info(`Retrying in ${delay}s...`);
    await new Promise((r) => setTimeout(r, delay * 1000));
    return sendWithRetry(toEmail, subject, htmlContent, attempt + 1);
  }

  log.error(`All email delivery attempts failed for ${toEmail}`);
  return result;
}

// ── Public API ────────────────────────────────────────────────────────

interface SendEmailOptions {
  to: string;
  subject: string;
  htmlBody: string;
}

interface SendTemplateOptions {
  to: string;
  userName?: string;
}

async function sendEmail({ to, subject, htmlBody }: SendEmailOptions): Promise<EmailDeliveryResult> {
  const validation = validateEmail(to);
  if (!validation.valid) {
    return {
      success: false,
      errorMessage: `Email validation failed: ${validation.error}`,
      statusCode: 400,
      timestamp: new Date(),
      retryCount: 0,
    };
  }

  if (!SMTP2GO_API_KEY) {
    log.warn(`SMTP2GO_API_KEY not configured. Would send to ${to}`);
    return {
      success: true,
      errorMessage: 'Email service not configured (dry-run mode)',
      statusCode: 0,
      timestamp: new Date(),
      retryCount: 0,
    };
  }

  return sendWithRetry(to, subject, htmlBody);
}

async function sendVerificationEmail({
  to,
  userName,
  verificationUrl,
}: SendTemplateOptions & { verificationUrl: string }): Promise<EmailDeliveryResult> {
  const subject = 'Verify Your Email - LightSerp';
  const htmlBody = getVerificationTemplate(verificationUrl, userName);
  return sendEmail({ to, subject, htmlBody });
}

async function sendPasswordResetEmail({
  to,
  userName,
  resetUrl,
}: SendTemplateOptions & { resetUrl: string }): Promise<EmailDeliveryResult> {
  const subject = 'Reset Your Password - LightSerp';
  const htmlBody = getPasswordResetTemplate(resetUrl, userName);
  return sendEmail({ to, subject, htmlBody });
}

async function sendWelcomeEmail({
  to,
  userName,
}: SendTemplateOptions & { userName: string }): Promise<EmailDeliveryResult> {
  const subject = 'Welcome to LightSerp!';
  const htmlBody = getWelcomeTemplate(userName, to);
  return sendEmail({ to, subject, htmlBody });
}

function isConfigured(): boolean {
  return !!SMTP2GO_API_KEY;
}

// ── Export ────────────────────────────────────────────────────────────

export {
  sendEmail,
  sendVerificationEmail,
  sendPasswordResetEmail,
  sendWelcomeEmail,
  isConfigured,
  validateEmail,
  type EmailDeliveryResult,
  type SendEmailOptions,
};
